docs/man/man1/gdal_pansharpen.1
.TH "gdal_pansharpen" 1 "Mon Nov 20 2017" "GDAL" \" -*- nroff -*-
.ad l
.nh
.SH NAME
gdal_pansharpen \- Perform a pansharpen operation\&.
.PP
(Since GDAL 2\&.1)
.SH "SYNOPSIS"
.PP
.PP
.nf
gdal_pansharpen [--help-general] pan_dataset {spectral_dataset[,band=num]}+ out_dataset
                [-of format] [-b band]* [-w weight_val]*
                [-r {nearest,bilinear,cubic,cubicspline,lanczos,average}]
                [-threads {ALL_CPUS|number}] [-bitdepth val] [-nodata val]
                [-spat_adjust {union,intersection,none,nonewithoutwarning}]
                [-co NAME=VALUE]* [-q]
.fi
.PP
.SH "DESCRIPTION"
.PP
The gdal_pansharpen\&.py script performs a pan-sharpening operation\&. It can create a 'classic' output dataset (such as GeoTIFF), or a VRT dataset describing the pan-sharpening operation\&.
.PP
More details can be found in the \fCVRT tutorial\fP\&.
.PP
.IP "\fB\fB-of\fP \fIformat\fP:\fP" 1c
Select the output format\&. The default is GeoTIFF (GTiff)\&. 'VRT' can also be used\&. Use the short format name\&. 
.IP "\fB\fB-b\fP \fIband\fP:\fP" 1c
Select band \fIband\fP from the input spectral bands for output\&. Bands are numbered from 1 in the order spectral bands are specified\&. Multiple \fB-b\fP switches may be used\&. When no -b switch is used, all input spectral bands are set for output\&. 
.IP "\fB\fB-w\fP \fIweight_val\fP:\fP" 1c
Specify a weight for the computation of the pseudo panchromatic value\&. There must be as many -w switches as input spectral bands\&. 
.IP "\fB\fB-r\fP \fI{nearest,bilinear,cubic (default),cubicspline,lanczos,average}\fP:\fP" 1c
Select a resampling algorithm\&. 
.IP "\fB\fB-threads\fP \fI{ALL_CPUS,number}\fP:\fP" 1c
Specify number of threads to use to do the resampling and pan-sharpening itself\&. Can be an integer number or ALL_CPUS\&. 
.IP "\fB\fB-bitdepth\fP \fIval\fP:\fP" 1c
Specify the bit depth of the panchromatic and spectral bands (e\&.g\&. 12)\&. If not specified, the NBITS metadata item from the panchromatic band will be used if it exists\&. 
.IP "\fB\fB-nodata\fP \fIval\fP:\fP" 1c
Specify nodata value for bands\&. Used for the resampling and pan-sharpening computation itself\&. If not set, deduced from the input bands, provided they have a consistent setting\&. 
.IP "\fB\fB-spat_adjust\fP \fI{union (default),intersection,none,nonewithoutwarning}\fP:\fP" 1c
Select behaviour when bands have not the same extent\&. See \fISpatialExtentAdjustment\fP documentation in \fCVRT tutorial\fP 
.IP "\fB\fB-co\fP \fI'NAME=VALUE'\fP:\fP" 1c
Passes a creation option to the output format driver\&. Multiple \fB-co\fP options may be listed\&. See \fCformat specific documentation for legal creation options for each format\fP\&. 
.IP "\fB\fB-q\fP:\fP" 1c
Suppress progress monitor and other non-error output\&. 
.IP "\fB\fIpan_dataset\fP\fP" 1c
Dataset with panchromatic band (first band will be used)\&. 
.IP "\fB\fIspectral_dataset[,band=num]\fP\fP" 1c
Dataset with one or several spectral bands\&. If the band option is not specified, all bands of the datasets are taken into account\&. Otherwise, only the specified (num)th band\&. The same dataset can be repeated several times\&. 
.IP "\fB\fIout_dataset\fP\fP" 1c
Output dataset 
.PP
.PP
Bands should be in the same projection\&.
.SH "EXAMPLE"
.PP
With spectral bands in a single dataset : 
.PP
.nf
gdal_pansharpen.py panchro.tif multispectral.tif pansharpened_out.tif

.fi
.PP
.PP
With a few spectral bands from a single dataset, reordered : 
.PP
.nf
gdal_pansharpen.py panchro.tif multispectral.tif,band=3 multispectral.tif,band=2 multispectral.tif,band=1 pansharpened_out.tif

.fi
.PP
.PP
With spectral bands in several datasets : 
.PP
.nf
gdal_pansharpen.py panchro.tif band1.tif band2.tif band3.tif pansharpened_out.tif

.fi
.PP
.PP
Specify weights: 
.PP
.nf
gdal_pansharpen.py -w 0.7 -w 0.2 -w 0.1 panchro.tif multispectral.tif pansharpened_out.tif

.fi
.PP
.PP
Specify RGB bands from a RGBNir multispectral dataset while computing the pseudo panchromatic intensity on the 4 RGBNir bands: 
.PP
.nf
gdal_pansharpen.py -b 1 -b 2 -b 3 panchro.tif rgbnir.tif pansharpened_out.tif

.fi
.PP
.SH "AUTHORS"
.PP
Even Rouault <even\&.rouault at spatialys\&.com> 


bindings/python/scripts/gdal_pansharpen.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
###############################################################################
# $Id: gdal_pansharpen.py 29568 2015-07-23 14:33:44Z rouault $
#
#  Project:  GDAL scripts
#  Purpose:  Perform a pansharpening operation
#  Author:   Even Rouault <even.rouault at spatialys.com>
#
###############################################################################
#  Copyright (c) 2015, Even Rouault <even.rouault at spatialys.com>
#
#  Permission is hereby granted, free of charge, to any person obtaining a
#  copy of this software and associated documentation files (the "Software"),
#  to deal in the Software without restriction, including without limitation
#  the rights to use, copy, modify, merge, publish, distribute, sublicense,
#  and/or sell copies of the Software, and to permit persons to whom the
#  Software is furnished to do so, subject to the following conditions:
#
#  The above copyright notice and this permission notice shall be included
#  in all copies or substantial portions of the Software.
#
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
#  OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
#  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
###############################################################################

import os
import sys
from osgeo import gdal

def Usage():
    print('Usage: gdal_pansharpen [--help-general] pan_dataset {spectral_dataset[,band=num]}+ out_dataset')
    print('                       [-of format] [-b band]* [-w weight]*')
    print('                       [-r {nearest,bilinear,cubic,cubicspline,lanczos,average}]')
    print('                       [-threads {ALL_CPUS|number}] [-bitdepth val] [-nodata val]')
    print('                       [-spat_adjust {union,intersection,none,nonewithoutwarning}]')
    print('                       [-verbose_vrt] [-co NAME=VALUE]* [-q]')
    print('')
    print('Create a dataset resulting from a pansharpening operation.')
    return -1

def gdal_pansharpen(argv):

    argv = gdal.GeneralCmdLineProcessor( argv )
    if argv is None:
        return -1

    pan_name = None
    last_name = None
    spectral_ds = []
    spectral_bands = []
    out_name = None
    bands = []
    weights = []
    format = 'GTiff'
    creation_options = []
    callback = gdal.TermProgress
    resampling = None
    spat_adjust = None
    verbose_vrt = False
    num_threads = None
    bitdepth = None
    nodata = None

    i = 1
    argc = len(argv)
    while i < argc:
        if argv[i] == '-of' and i < len(argv)-1:
            format = argv[i+1]
            i = i + 1
        elif argv[i] == '-r' and i < len(argv)-1:
            resampling = argv[i+1]
            i = i + 1
        elif argv[i] == '-spat_adjust' and i < len(argv)-1:
            spat_adjust = argv[i+1]
            i = i + 1
        elif argv[i] == '-b' and i < len(argv)-1:
            bands.append(int(argv[i+1]))
            i = i + 1
        elif argv[i] == '-w' and i < len(argv)-1:
            weights.append(float(argv[i+1]))
            i = i + 1
        elif argv[i] == '-co' and i < len(argv)-1:
            creation_options.append(argv[i+1])
            i = i + 1
        elif argv[i] == '-threads' and i < len(argv)-1:
            num_threads = argv[i+1]
            i = i + 1
        elif argv[i] == '-bitdepth' and i < len(argv)-1:
            bitdepth = argv[i+1]
            i = i + 1
        elif argv[i] == '-nodata' and i < len(argv)-1:
            nodata = argv[i+1]
            i = i + 1
        elif argv[i] == '-q':
            callback = None
        elif argv[i] == '-verbose_vrt':
            verbose_vrt = True
        elif argv[i][0] == '-':
            sys.stderr.write('Unrecognized option : %s\n' % argv[i])
            return Usage()
        elif pan_name is None:
            pan_name = argv[i]
            pan_ds = gdal.Open(pan_name)
            if pan_ds is None:
                return 1
        else:
            if last_name is not None:
                pos = last_name.find(',band=')
                if pos > 0:
                    spectral_name = last_name[0:pos]
                    ds = gdal.Open(spectral_name)
                    if ds is None:
                        return 1
                    band_num = int(last_name[pos+len(',band='):])
                    band = ds.GetRasterBand(band_num)
                    spectral_ds.append(ds)
                    spectral_bands.append(band)
                else:
                    spectral_name = last_name
                    ds = gdal.Open(spectral_name)
                    if ds is None:
                        return 1
                    for j in range(ds.RasterCount):
                        spectral_ds.append(ds)
                        spectral_bands.append(ds.GetRasterBand(j+1))

            last_name = argv[i]

        i = i + 1

    if pan_name is None or len(spectral_bands) == 0:
        return Usage()
    out_name = last_name

    if len(bands) == 0:
        bands = [ j+1 for j in range(len(spectral_bands)) ]
    else:
        for i in range(len(bands)):
            if bands[i] < 0 or bands[i] > len(spectral_bands):
                print('Invalid band number in -b: %d' % bands[i])
                return 1

    if len(weights) != 0 and len(weights) != len(spectral_bands):
        print('There must be as many -w values specified as input spectral bands')
        return 1

    vrt_xml = """<VRTDataset subClass="VRTPansharpenedDataset">\n"""
    if bands != [ j+1 for j in range(len(spectral_bands)) ]:
        for i in range(len(bands)):
            band = spectral_bands[bands[i]-1]
            datatype = gdal.GetDataTypeName(band.DataType)
            colorname = gdal.GetColorInterpretationName(band.GetColorInterpretation())
            vrt_xml += """  <VRTRasterBand dataType="%s" band="%d" subClass="VRTPansharpenedRasterBand">
      <ColorInterp>%s</ColorInterp>
  </VRTRasterBand>\n""" % (datatype, i+1, colorname)

    vrt_xml += """  <PansharpeningOptions>\n"""

    if len(weights) != 0:
        vrt_xml += """      <AlgorithmOptions>\n"""
        vrt_xml += """        <Weights>"""
        for i in range(len(weights)):
            if i > 0: vrt_xml += ","
            vrt_xml += "%.16g" % weights[i]
        vrt_xml += "</Weights>\n"
        vrt_xml += """      </AlgorithmOptions>\n"""

    if resampling is not None:
        vrt_xml += '      <Resampling>%s</Resampling>\n' % resampling

    if num_threads is not None:
        vrt_xml += '      <NumThreads>%s</NumThreads>\n' % num_threads

    if bitdepth is not None:
        vrt_xml += '      <BitDepth>%s</BitDepth>\n' % bitdepth

    if nodata is not None:
        vrt_xml += '      <NoData>%s</NoData>\n' % nodata

    if spat_adjust is not None:
        vrt_xml += '      <SpatialExtentAdjustment>%s</SpatialExtentAdjustment>\n' % spat_adjust

    pan_relative='0'
    if format.upper() == 'VRT':
        if not os.path.isabs(pan_name):
            pan_relative='1'
            pan_name = os.path.relpath(pan_name, os.path.dirname(out_name))

    vrt_xml += """    <PanchroBand>
      <SourceFilename relativeToVRT="%s">%s</SourceFilename>
      <SourceBand>1</SourceBand>
    </PanchroBand>\n""" % (pan_relative, pan_name)

    for i in range(len(spectral_bands)):
        dstband = ''
        for j in range(len(bands)):
            if i + 1 == bands[j]:
                dstband = ' dstBand="%d"' % (j+1)
                break

        ms_relative='0'
        ms_name = spectral_ds[i].GetDescription()
        if format.upper() == 'VRT':
            if not os.path.isabs(ms_name):
                ms_relative='1'
                ms_name = os.path.relpath(ms_name, os.path.dirname(out_name))

        vrt_xml += """    <SpectralBand%s>
      <SourceFilename relativeToVRT="%s">%s</SourceFilename>
      <SourceBand>%d</SourceBand>
    </SpectralBand>\n""" % (dstband, ms_relative, ms_name, spectral_bands[i].GetBand())

    vrt_xml += """  </PansharpeningOptions>\n"""
    vrt_xml += """</VRTDataset>\n"""

    if format.upper() == 'VRT':
        f = gdal.VSIFOpenL(out_name, 'wb')
        if f is None:
            print('Cannot create %s' % out_name)
            return 1
        gdal.VSIFWriteL(vrt_xml, 1, len(vrt_xml), f)
        gdal.VSIFCloseL(f)
        if verbose_vrt:
            vrt_ds = gdal.Open(out_name, gdal.GA_Update)
            vrt_ds.SetMetadata(vrt_ds.GetMetadata())
        else:
            vrt_ds = gdal.Open(out_name)
        if vrt_ds is None:
            return 1

        return 0

    vrt_ds = gdal.Open(vrt_xml)
    out_ds = gdal.GetDriverByName(format).CreateCopy(out_name, vrt_ds, 0, creation_options, callback = callback)
    if out_ds is None:
        return 1
    return 0

def main():
    return gdal_pansharpen(sys.argv)

if __name__ == '__main__':
    sys.exit(gdal_pansharpen(sys.argv))
